From 1181d24c20b79d6a9de47daa3be8bcd6c0a2bbe5 Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Mon, 24 Nov 2014 16:50:27 -0500
Subject: [powerpc] xmon: le - Fix endiannes issue in RTAS call from xmon

Message-id: <20141124165027.21465.26552.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 100859
O-Subject: [PATCH RHEL7.1 BZ1160650] pseries/le: Fix endiannes issue in RTAS call from xmon
Bugzilla: 1160650
RH-Acked-by: David Gibson <dgibson@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1160650

Description:
------------
On pseries system (LPAR) xmon failed to enter when running in LE mode, system
is hunging. Inititating xmon will lead to such an output on the console:

SysRq : Entering xmon
cpu 0x15: Vector: 0  at [c0000003f39ffb10]
    pc: c00000000007ed7c: sysrq_handle_xmon+0x5c/0x70
    lr: c00000000007ed7c: sysrq_handle_xmon+0x5c/0x70
    sp: c0000003f39ffc70
   msr: 8000000000009033
  current = 0xc0000003fafa7180
  paca    = 0xc000000007d75e80  softe: 0  irq_happened: 0x01
    pid   = 14617, comm = bash
Bad kernel stack pointer fafb4b0 at eca7cc4
cpu 0x15: Vector: 300 (Data Access) at [c000000007f07d40]
    pc: 000000000eca7cc4
    lr: 000000000eca7c44
    sp: fafb4b0
   msr: 8000000000001000
   dar: 10000000
 dsisr: 42000000
  current = 0xc0000003fafa7180
  paca    = 0xc000000007d75e80  softe: 0  irq_happened: 0x01
    pid   = 14617, comm = bash
cpu 0x15: Exception 300 (Data Access) in xmon, returning to main loop
xmon: WARNING: bad recursive fault on cpu 0x15

The root cause is that xmon is calling RTAS to turn off the surveillance
when entering xmon, and RTAS is requiring big endian parameters.

This patch is byte swapping the RTAS arguments when running in LE mode.

Signed-off-by: Laurent Dufour <ldufour@linux.vnet.ibm.com>

Brew:
-----
BE: https://brewweb.devel.redhat.com/taskinfo?taskID=8288783
LE: https://brewweb.devel.redhat.com/taskinfo?taskID=8288811

Upstream:
---------
https://git.kernel.org/cgit/linux/kernel/git/mpe/linux.git/commit/?h=fixes&id=3b8a3c01096925a824ed3272601082289d9c23a5

Posting now to try to make the beta cutoff if it isn't too late.

Test Status:
------------
Tested successfully by me on both BE and LE power system (ibm-p8-01-lp3).
on power LE

with the patch we are now able to enter and exit xmon successfully.

[root@ibm-p8-01-lp3 ~]# echo x > /proc/sysrq-trigger
[  105.356673] SysRq : Entering xmon
cpu 0x1: Vector: 0  at [c0000003eea7fb80]
    pc: c000000000561560: write_sysrq_trigger+0x120/0x260
    lr: c000000000561560: write_sysrq_trigger+0x120/0x260
    sp: c0000003eea7fce0
   msr: 8000000000009033
  current = 0xc0000003eccc6130
  paca    = 0xc000000007df0380   softe: 0        irq_happened: 0x00
    pid   = 11970, comm = bash
enter ? for help
[c0000003eea7fd90] c00000000037d7d4 proc_reg_write+0x84/0x120
[c0000003eea7fdd0] c0000000002d51e8 SyS_write+0x148/0x390
[c0000003eea7fe30] c00000000000a0fc syscall_exit+0x0/0x7c
--- Exception: c00 (System Call) at 00003fff7b122980
SP (3fffef84c940) is in userspace

1:mon> dl
[    0.000000] Reserving 512MB of memory at 128MB for crashkernel (System RAM: 16384MB)
[    0.000000] Allocated 1835008 bytes for 2048 pacas at c000000007df0000
[    0.000000] Using pSeries machine description

Verified that the patched kernel still works on BE power.

[root@ibm-p8-01-lp3 ~]# echo x > /proc/sysrq-trigger
[   65.327926] SysRq : Entering xmon
cpu 0x1: Vector: 0  at [c0000003ade2fad0]
    pc: c00000000055ec90: .write_sysrq_trigger+0x120/0x250
    lr: c00000000055ec90: .write_sysrq_trigger+0x120/0x250
    sp: c0000003ade2fc30
   msr: 8000000000009032
  current = 0xc0000003b4c06130
  paca    = 0xc000000007b30900   softe: 0        irq_happened: 0x00
    pid   = 2896, comm = bash
enter ? for help
[c0000003ade2fcf0] c00000000037e510 .proc_reg_write+0x80/0x120
[c0000003ade2fd80] c0000000002d7774 .SyS_write+0x144/0x390
[c0000003ade2fe30] c00000000000a17c syscall_exit+0x0/0x7c
--- Exception: c00 (System Call) at 00003fff8c003570
SP (3fffcabea440) is in userspace
1:mon> dl
[    0.000000] Reserving 512MB of memory at 128MB for crashkernel (System RAM: 16384MB)
[    0.000000] Allocated 4718592 bytes for 2048 pacas at c000000007b30000
[    0.000000] Using pSeries machine description

---------------------------------------------------------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/xmon/xmon.c b/arch/powerpc/xmon/xmon.c
index d199bfa..c6c497b 100644
--- a/arch/powerpc/xmon/xmon.c
+++ b/arch/powerpc/xmon/xmon.c
@@ -292,10 +292,10 @@ static inline void disable_surveillance(void)
  args.token = rtas_token("set-indicator");
  if (args.token == RTAS_UNKNOWN_SERVICE)
   return;
- args.nargs = 3;
- args.nret = 1;
+ args.nargs = cpu_to_be32(3);
+ args.nret = cpu_to_be32(1);
  args.rets = &args.args[3];
- args.args[0] = SURVEILLANCE_TOKEN;
+ args.args[0] = cpu_to_be32(SURVEILLANCE_TOKEN);
  args.args[1] = 0;
  args.args[2] = 0;
  enter_rtas(__pa(&args));
-- 
1.7.1